File: old/sorting/quick_sort/lomutu_quick_select/lomutu_quick_select.py
def lomutu_partition(arr, start, end):
    left = start
    pivot = arr[end]

    for i in range(start, end):
        if arr[i] < pivot:
            arr[i], arr[left] = arr[left], arr[i]
            left += 1

    arr[end] = arr[left]
    arr[left] = pivot

    return arr, left


def lomuto_quick_select(arr, start, end, k):
    if start >= end:
        return arr[k]

    arr, partition_index = lomutu_partition(arr, start, end)

    if partition_index == k:
        return arr[partition_index]
    elif partition_index > k:
        return lomuto_quick_select(arr, start, partition_index - 1, k)
    else:
        return lomuto_quick_select(arr, partition_index + 1, end, k)

    return arr[k]


def main():
    input = [7, 10, 4, 3, 20, 15]
    nth_smallest_element_index = 4
    output = lomuto_quick_select(input, 0, len(input) - 1, nth_smallest_element_index)
    print(output)


main()

File: old/sorting/quick_sort/lomutu_quick_select/lomutu_quick_select.cpp
#include <iostream>
#include <vector>

using namespace std;

int lomutuPartition(vector<int>& arr, int start, int end) { 
    int left = start;
    int pivot = arr[end];

    for (int i = start; i < end; i++) {
        if (arr[i] < pivot) {
            swap(arr[i], arr[left]);
            left += 1;
        }
    }

    arr[end] = arr[left];
    arr[left] = pivot;

    return left;
}

int lomutuQuickSelect(vector<int>& arr, int start, int end, int& k) {
    if (start >= end) {
        return arr[k];
    }

    int partitionIndex = lomutuPartition(arr, start, end);
    if (partitionIndex == k) {
        return arr[partitionIndex];
    } else if (partitionIndex > k) {
        return lomutuQuickSelect(arr, start, partitionIndex - 1, k);
    } else {
        return lomutuQuickSelect(arr, partitionIndex + 1, end, k);
    }

    return arr[partitionIndex];
}

int main() {
    vector<int> input = {7, 10, 4, 3, 20, 15};
    int nth_smallest_element_index = 4;
    int result = lomutuQuickSelect(input, 0, input.size() - 1, nth_smallest_element_index);
    cout << result << endl;
    return 0;
}
